### PR TITLE
pmi: add support for pmi2

### PR DESCRIPTION
**Pull request by hppritcha**
This commit adds support for using pmi2 and hence
the ability to use the Cray PMI.  Cray PMI is weird in
several ways:

1) the PMI_KVX functions only work when craypich is launched
   using hydra (only supported on single node)
2) Cray has PMI2 but doesn't install a libpmi2
3) Cray has PMI2 but doesn't define PMI2_SUCCESS.

Verified using a modified ofi-transport.c that some of the
perf test work using socket provider on cray.

More work is need to the ofi layer to get the GNI provider
to work with sandia shmem.

@kseager 
@jdinan 

Signed-off-by: Howard Pritchard <howardp@lanl.gov>

### CODE REVIEW

**Comment by jdinan** [comment]
Could you look into the build breakages that Travis is reporting?

Also, can we move the USE_PMI1/2 guards into the Makefile rather than in the runtime-pmi1/2.c files?

**Comment by hppritcha** [comment]
I need to do a configury fix for the case where one is using the built-in simple pmi. Working on it.

**Comment by hppritcha** [comment]
@jdinan check now

**Comment by jdinan** [comment]
Thanks for the updates.  One more question -- I'm not too familiar with the OMPI build system, what was the reason for replacling orte_check_pmi.m4 with opal_...?  Was the old file just out dated?  I assume that they are functionally equivalent?

**Comment by hppritcha** [comment]
it was out of date.  we switched to using opal_check_pmi.m4 as part of the pmix work.

**Comment by hppritcha** [comment]
I didn't want to remove the old orte_check_pmi.m4 yet.

File: src/runtime-pmi2.c [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+ * Run-time support for the built-in runtime that is part of the
+ * shared memory implementation Portals
+ */
+
+#include "config.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#if defined(PMI_SLURM)
```
Comment (kseager) [inline src/runtime-pmi2.c]: you aren't defining this in your m4 like we were in orte...m4 and this causes build issues with the slurm+pmi2 combo

Comment (hppritcha) [inline src/runtime-pmi2.c]: okay, fixing.

**Comment by hppritcha** [comment]
@kseager now try on a slurm/pmi2 installation.  Thanks.

**Comment by hppritcha** [comment]
Fixed a problem with setting the max allowable value len for pmi2.

**Comment by hppritcha** [comment]
added @kseager 's commits to allow for explicit pmi1 configury selection as well as to fix support
for portals4 pmi.

**Pull request merged**